File: CrackingTheCodeInterview/Array&Strings/C++/OneAway.cpp
#include <bits/stdc++.h>
#if !defined(True)
#define True  (1==1)
#endif
#if !defined(False)
#define False (!True)
#endif
using namespace std;

bool addCharDistance(string A, string B){
	size_t lenA = A.length(), lenB = B.length(), i = 0, j = 0, dif = 0;		// O(1)
	while(i < lenA && j < lenB){									        // O(A)
		if(dif > 1) return False;										    	// O(1)
		if(A[i] != B[j]){i++; dif++;}											// O(1)
		else{i++; j++;}															// O(1)
	}
	dif += lenA - i;														// O(1)
	return (dif <= 1)? True : False; 										// O(1)
}

bool changeCharDistance(string A, string B){
	size_t lenA = A.length(), lenB = B.length(), i = 0, j = 0, dif = 0;		// O(1)
	while(i < lenA && j < lenB)										        // O(N)
		if(dif > 1) return False;										    	// O(1)
		else if(A[i++] != B[j++]) dif++;										// O(1)
	return (dif <= 1)? True : False; 										// O(1)
}

bool oneAwaySimple(string A, string B){
	size_t lenA = A.length(), lenB = B.length(), i = 0, j = 0, dif = 0;		// O(1)
	while(i < lenA && j < lenB){									        // O(A)
		if(dif > 1) return False;										    	// O(1)
		if(A[i] != B[j])
			if(lenA > lenB){i++; dif++;}										// O(1)
			else{i++; j++; dif++;}												// O(1)
		else{i++; j++;}															// O(1)
	}
	dif += lenA - i;														// O(1)
	return (dif <= 1)? True : False; 										// O(1)
}

bool oneAway(string A, string B){
	size_t lenA = A.length(), lenB = B.length(), i = 0, j = 0;				// O(1)
	int dif = lenA - lenB;													// O(1)
	if(abs(dif) > 1) return False;											// O(1)
	else if(dif < 0) return oneAwaySimple(B, A);							// O(N) N = maxLength(lenA, lenB)
	return oneAwaySimple(A, B);												// O(N) N = maxLength(lenA, lenB)
	// if(dif == 0) return changeCharDistance(A, B);						// O(N)
	// else if(dif < 0) return addCharDistance(B, A)						// O(N) N = maxLength(lenA, lenB)
	// return addCharDistance(A, B);										// O(N) N = maxLength(lenA, lenB)
}

int main(int argc, char const *argv[]){
	ios_base::sync_with_stdio(false);
	string s = "aDb";
	string t = "adb";
	cout << oneAway(s, t) << "\n";
	cout << oneAway("bcde", "abcde") << "\n";
	cout << oneAway("pale", "ple") << "\n";
	cout << oneAway("pales", "pale") << "\n";
	cout << oneAway("pale", "bale") << "\n";
	cout << oneAway("pale", "bake") << "\n";
	cout << oneAway("pal", "palee") << "\n";
	return 0;
}

File: CrackingTheCodeInterview/Array&Strings/Python3/OneAway.py
def addCharDistance(stringA, stringB):
	lenA, lenB, i, j, dif = len(stringA), len(stringB), 0, 0, 0		# O(1)
	while i < lenA and j < lenB:									# O(A)
		if dif > 1: return False										# O(1)
		if stringA[i] != stringB[j]:									# O(1)
			i += 1														# O(1)
			dif += 1													# O(1)
		else:															# O(1)
			i += 1														# O(1)
			j += 1
	dif += lenA - i													# O(1)
	return True if dif <= 1 else False 								# O(1)

def changeCharDistance(stringA, stringB):
	lenA, lenB, i, j, dif = len(stringA), len(stringB), 0, 0, 0		# O(1)
	while i < lenA and j < lenB:									# O(N)
		if dif > 1: return False										# O(1)
		if stringA[i] != stringB[j]: dif += 1							# O(1)					
		i += 1															# O(1)
		j += 1															# O(1)
	return True if dif <= 1 else False									# O(1)

def oneAwaySimple(A, B):
	lenA = len(A)					# O(1)
	lenB = len(B)					# O(1)
	i, j, dif = 0, 0, 0				# O(1)
	while i < lenA and j < lenB:	# O(A)
		if dif > 1: return False				# O(1)
		if A[i] != B[j]:						# O(1)
			if lenA > lenB:						# O(1)
				i, dif = i + 1, dif +1 				# O(1)
			else:								# O(1)
				i, j, dif = i + 1, j + 1, dif +1	# O(1)
		else:									# O(1)
			i, j = i + 1, j + 1						# O(1)
	dif += lenA - i;						# O(1)
	return True if dif <= 1 else False; 	# O(1)


def oneAway(stringA, stringB):
	lenA, lenB, i, j = len(stringA), len(stringB), 0, 0				# O(1)
	dif = lenA - lenB												# O(1)
	if abs(dif) > 1: return False									# O(1)
	elif dif < 0: 	return oneAwaySimple(stringB, stringA)		# O(N) N = maxLength(lenA, lenB)
	return oneAwaySimple(stringA, stringB)						# O(N) N = maxLength(lenA, lenB)
	# if dif == 0: return changeCharDistance(stringA, stringB)		# O(N)
	# elif dif < 0: 	return addCharDistance(stringB, stringA)		# O(N) N = maxLength(lenA, lenB)
	# return addCharDistance(stringA, stringB)						# O(N) N = maxLength(lenA, lenB)

s = "aDb"
t = "adb"

print(oneAway(s, t))
print(oneAway("bcde", "abcde"))
print(oneAway('pale', 'ple'))
print(oneAway('pales', 'pale'))
print(oneAway('pale', 'bale'))
print(oneAway('pale', 'bake'))
print(oneAway('pal', 'palee'))
